framework/src/service/ServiceReferenceBasePrivate.cpp

```diff
@@ -103,64 +114,51 @@ InterfaceMapConstPtr ServiceReferenceBasePrivate::GetPrototypeService(const Bund
 
 std::shared_ptr<void> ServiceReferenceBasePrivate::GetService(BundlePrivate* bundle)
 {
-  auto s = ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
-  if (!s)
-  {
-    registration->Lock(), --registration->dependents[bundle];
-  }
-  return s;
+  return ExtractInterface(GetServiceInterfaceMap(bundle), interfaceId);
 }
 
 InterfaceMapConstPtr ServiceReferenceBasePrivate::GetServiceInterfaceMap(BundlePrivate* bundle)
 {
   InterfaceMapConstPtr s;
   if (!registration->available) return s;
-
   std::shared_ptr<ServiceFactory> serviceFactory;
-  int count = 0;
 
   {
     auto l = registration->Lock(); US_UNUSED(l);
     if (!registration->available) return s;
     serviceFactory = std::static_pointer_cast<ServiceFactory>(
           registration->GetService_unlocked("org.cppmicroservices.factory"));
-    count = registration->dependents[bundle];
-  }
 
-  if (!serviceFactory)
-  {
-    auto l = registration->Lock(); US_UNUSED(l);
-          s = registration->service;
-    if (s && !s->empty()) ++registration->dependents[bundle];
-  }
-  else
-  {
-    if (count == 0)
-    {
-      s = GetServiceFromFactory(bundle, serviceFactory);
+    registration->dependents.insert(std::make_pair(bundle, 0));
 
-      auto l = registration->Lock(); US_UNUSED(l);
-      if (registration->dependents[bundle] == 0)
-      {
-        registration->bundleServiceInstance.insert(std::make_pair(bundle, s));
-      }
-      else
+    if (!serviceFactory)
+    {
+      s = registration->service;
+      if (s && !s->empty())
       {
-        // There was a race and we now have one instance too much. Return the
-        // already produced instance and ignore the additional one.
-        s = registration->bundleServiceInstance[bundle];
+        ++registration->dependents.at(bundle);
       }
-
-      if (s && !s->empty()) ++registration->dependents[bundle];
-    }
-    else
-    {
-      auto l = registration->Lock(); US_UNUSED(l);
-      // return the already produced instance
-      s = registration->bundleServiceInstance[bundle];
-      if (s && !s->empty()) ++registration->dependents[bundle];
+      return s;
     }
   }
+
+  // Calling into a service factory could cause re-entrancy into the
+  // framework and even, theoretically, into this function. Ensuring
+  // we don't hold a lock while calling into the service factory eliminates
+  // the possibility of a deadlock. It does not however eliminate the
+  // possbilituy of infinite recursion.
+  s = GetServiceFromFactory(bundle, serviceFactory);
```

Comment:
My goal was to simplify the code and reduce the number of conditional branches (at the cost of multiple ```GetServiceFromFactory``` calls).
 
It is possible to avoid calling ```GetServiceFromFactory``` each time and still have no race condition. The drawback is that duplicate code would be required in both locked regions of code. 
Once the code exits the first lock, no assumptions can be made about the state of ```registration->dependents``` and ```registration->bundleServiceInstance``` when entering the second lock. Both locked regions of code need to do similar, if not the same, checks after acquiring the lock.

Is this drawback (i.e. duplicate code) acceptable?